### PR TITLE
Winobias reader

### PR DESCRIPTION
**Pull request by DeNeutoy**
This dataset reader reads MarkY's Winobias dataset, for evaluating bias in coref models. Turn's out our coref model is pretty biased, but thanks to Mark Y we can fix it. This is a first step towards that.

Run with:
```
python allennlp/run.py evaluate \
--evaluation-data-file PATH_TO_WINOBIAS_FILE \
-o '{"dataset_reader": {"type": "winobias", "token_indexers": {"tokens": {"type": "single_id", "lowercase_tokens": true}, "token_characters": {"type": "characters"}}}}' \
https://s3-us-west-2.amazonaws.com/allennlp/models/coref-model-2018.02.05.tar.gz
```

Type 1 example: 
[entity1] [interacts with] [entity2] [conjunction] [pronoun] [circumstances].
http://demo.allennlp.org/coreference-resolution/NjE2OTg=
vs 
http://demo.allennlp.org/coreference-resolution/NjE2OTk=

Type 2 example: 
[entity1] [interacts with] [entity2] and then [interacts with] [pronoun] for [circumstances].
http://demo.allennlp.org/coreference-resolution/NjE3MDI=
vs
http://demo.allennlp.org/coreference-resolution/NjE3MDM=

Type 1 F1 Score difference: 12pts, 64.8 vs 53.0
Type 2 F1 Score difference: 7pts, 77.4 vs 69.8

### CODE REVIEW

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    Winobias is a dataset to analyse the issue of gender bias in co-reference
+    resolution. It contains simple sentences with pro/anti stereotypical gender
+    associations with which to measure the bias of a coreference system trained
+    on another corpora.
+
+    Returns a list of ``Instances`` which have four fields: ``text``, a ``TextField``
+    containing the full sentence text, ``spans``, a ``ListField[SpanField]`` of inclusive start and
+    end indices for span candidates, and ``metadata``, a ``MetadataField`` that stores the instance's
+    original text. For data with gold cluster labels, we also include the original ``clusters``
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/coreference_resolution/winobias.py]: You include these in the `metadata`, right?

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+@DatasetReader.register("winobias")
+class WinobiasReader(DatasetReader):
+    """
+    TODO(Mark): Add paper reference.
+
+    Winobias is a dataset to analyse the issue of gender bias in co-reference
+    resolution. It contains simple sentences with pro/anti stereotypical gender
+    associations with which to measure the bias of a coreference system trained
+    on another corpora.
+
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/coreference_resolution/winobias.py]: It'd be nice to have a quick note on the input file format here.

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                 max_span_width: int,
+                 token_indexers: Dict[str, TokenIndexer] = None,
+                 lazy: bool = False) -> None:
+        super().__init__(lazy)
+        self._max_span_width = max_span_width
+        self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
+
+    @overrides
+    def _read(self, file_path: str):
+        # if `file_path` is a URL, redirect to the cache
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/coreference_resolution/winobias.py]: Formatting on the comment.  Though, the name `cached_path` is obvious enough, I'm not sure the comment is even necessary.  I'd probably even just inline the `cached_path` function call with `for sentence in open(cached_path(file_path), 'r'):`.

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+        for sentence in open(file_path, "r"):
+            tokens = sentence.strip().split(" ")
+            clusters: DefaultDict[int, List[Tuple[int, int]]] = collections.defaultdict(list)
+            words = []
+            inside_square_span = False
+            inside_round_span = False
+            for index, token in enumerate(tokens):
+                # Coreference is annotated using [square brackets]
+                # or (round brackets) around coreferent phrases.
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/coreference_resolution/winobias.py]: This logic looks like you're assuming you'll only ever have two coreference clusters in your input.  Is that right?  If so, that'd be another nice thing to note in the docstring above when you talk about file format.

File: tests/fixtures/coref/winobias.sample [hunk 5]
```diff
@@ -0,0 +1,2 @@
+The designer argued with [the developer] and slapped [her] in the face.
+[The salesperson] sold (some books) to the librarian because [she] was trying to sell (them).
```
Comment (matt-gardner) [inline tests/fixtures/coref/winobias.sample]: Newline at end of file.

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                if "[" in token and "]" in token:
+                    clusters[0].append((index, index))
+                elif "[" in token:
+                    clusters[0].append((index, index))
+                    inside_square_span = True
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/coreference_resolution/winobias.py]: Why do you need this flag?  Can't you just use the index when you get to `]`?  Also, looks like this logic won't work for nested annotations.  I assume you don't have those either?

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                    clusters[1].append((index, index))
+                    inside_round_span = True
+                elif inside_round_span or ")" in token:
+                    old_span = clusters[1][-1]
+                    clusters[1][-1] = (old_span[0], old_span[1] + 1)
+                    inside_round_span = False
+
+                if token.endswith("."):
+                    # Winobias is tokenised, but not for full stops.
+                    # We'll just special case them here.
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/coreference_resolution/winobias.py]: Does it already split things like "don't"?  And won't this behave poorly for things like "e.g."?

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    @overrides
+    def text_to_instance(self,  # type: ignore
+                         sentence: List[str],
+                         gold_clusters: Optional[List[List[Tuple[int, int]]]] = None) -> Instance:
+        # pylint: disable=arguments-differ
+        """
+        Parameters
+        ----------
+        sentences : ``List[str]``, required.
+            The sentence to analyse.
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/coreference_resolution/winobias.py]: ", already tokenized."

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                The id of the cluster which each possible span belongs to, or -1 if it does
+                 not belong to a cluster. As these labels have variable length (it depends on
+                 how many spans we are considering), we represent this a as a ``SequenceLabelField``
+                 with respect to the ``spans ``ListField``.
+        """
+        metadata: Dict[str, Any] = {"original_text": sentence}
+        if gold_clusters is not None:
+            metadata["clusters"] = gold_clusters
+
+        text_field = TextField([Token(word) for word in sentence], self._token_indexers)
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/coreference_resolution/winobias.py]: This probably won't interact well with a `Predictor`, which will probably use a tokenizer to process input.  You might want to have the conversion to `Tokens` in `_read`.

File: tests/data/dataset_readers/coreference/winobias_reader_test.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
+# pylint: disable=no-self-use,invalid-name
+from typing import List, Tuple
+
+import pytest
+
+from allennlp.data.dataset_readers import WinobiasReader
+from allennlp.common.util import ensure_list
+
+class TestWinobiasReader:
```
Comment (matt-gardner) [inline tests/data/dataset_readers/coreference/winobias_reader_test.py]: Inherit from our test case class?  I guess it doesn't matter here...

File: tests/data/dataset_readers/coreference/winobias_reader_test.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                                                               for i, x in gold_indices_with_ids]
+        assert gold_mentions_with_ids == [(['The', 'salesperson'], 0),
+                                          (['some', 'books'], 1),
+                                          (['she'], 0), (['them'], 1)]
+
+    def check_candidate_mentions_are_well_defined(self, span_starts, span_ends, text):
+        candidate_mentions = []
+        for start, end in zip(span_starts, span_ends):
+            # Spans are inclusive.
+            text_span = text[start: end + 1]
```
Comment (matt-gardner) [inline tests/data/dataset_readers/coreference/winobias_reader_test.py]: I think the right way to format this is `text[start:(end + 1)]`.  The spacing just makes me do a double-take if there are no parentheses.

File: allennlp/data/dataset_readers/coreference_resolution/winobias.py (hunk 7 above)
Comment (DeNeutoy) [reply allennlp/data/dataset_readers/coreference_resolution/winobias.py]: the sentences are so simple, they don't contain things like this. Sorry I should have made clear how simplistic this data is in the docstring.

Comment (DeNeutoy) [reply allennlp/data/dataset_readers/coreference_resolution/winobias.py]: And yes they already tokenise the few abbreviations they do have

**Comment by my89** [comment]
@DeNeutoy Jieyu, one of the collaborators on the project pointed out AllenNLP likely doesn't load some of the pre-trained vectors required for the WinoBias task (that would be available in glove) since they are not part of ConLL training/testing. This does seem to be borne out in that the performance of the AllenNLP the winobias data is down 3 points on type-1 and 10 points on type-2 relative to Kenton's system (which is likely why it is reported as much less biased than Kenton's system).

**Pull request merged**